op-batcher/batcher/espresso.go

```diff
@@ -256,8 +260,100 @@ func (l *BlockLoader) EnqueueBlocks(ctx context.Context, blocksToQueue inclusive
 			continue
 		}
 
-		l.lastQueuedBlock = &blockRef
+		l.queuedBlocks = append(l.queuedBlocks, blockRef)
+	}
+}
+
+type EnqueueBlockAction uint
+
+const (
+	ActionEnqueue = iota
+	ActionRetry
+	ActionReset
+)
+
+// This function is an analogue of `computeSyncActions` for Espresso batcher mode
+//
+// It computes the next block range to enqueue to Espresso based on new newSyncStatus and
+// does a number of checks to ensure consistency of the chain.
+//
+// If reorg is detected, empty range and ActionReset is returned.
+// If there isn't enough information or no blocks to load yet, empty range and ActionRetry is returned.
+func (l *BlockLoader) nextBlockRange(newSyncStatus *eth.SyncStatus) (inclusiveBlockRange, EnqueueBlockAction) {
+	if newSyncStatus.HeadL1 == (eth.L1BlockRef{}) {
+		// empty sync status
+		return inclusiveBlockRange{}, ActionRetry
+	}
+
+	if l.prevSyncStatus == nil {
+		l.prevSyncStatus = newSyncStatus
+	}
+
+	if newSyncStatus.CurrentL1.Number < l.prevSyncStatus.CurrentL1.Number {
+		// sequencer restarted and hasn't caught up yet
+		l.batcher.Log.Warn("sequencer currentL1 reversed", "new currentL1", newSyncStatus.CurrentL1.Number, "previous currentL1", l.prevSyncStatus.CurrentL1)
+		return inclusiveBlockRange{}, ActionRetry
+	}
+
+	var safeL2 eth.L2BlockRef
+	if l.batcher.Config.PreferLocalSafeL2 {
+		// This is preffered when running interop, but not yet enabled by default.
+		safeL2 = newSyncStatus.LocalSafeL2
+	} else {
+		safeL2 = newSyncStatus.SafeL2
+	}
+
+	// State empty, just enqueue all unsafe blocks
+	if len(l.queuedBlocks) == 0 {
+		return inclusiveBlockRange{safeL2.Number, newSyncStatus.UnsafeL2.Number}, ActionEnqueue
```

Comment:
Should not it be safeL2.Number+1 as by definition  safeL2.Number refers to a block that can be derived from L1?